File: 0146-lru-cache/0146-lru-cache.py
class LRUCache:
    d = dict()
    cap = 0
    li = list()

    def __init__(self, capacity: int):
        self.d = dict()
        self.li = list()
        self.cap = capacity

    def get(self, key: int) -> int:
        # print(self.d, self.li, self.cap)
        if key in self.d:
            val = self.d[key]
            self.li.remove(key)
            self.li.append(key)
            return val
        return -1

    def put(self, key: int, value: int) -> None:
        # print("put : ", self.d, self.li, self.cap)
        if key in self.d:
            self.li.remove(key)
            self.li.append(key)
            self.d[key] = value
            return
        if len(self.li)>=self.cap:
            first = self.li.pop(0)
            self.d.pop(first)
        self.li.append(key)
        self.d[key] = value
            

        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.cpp
class Node{
    public:
    int val;
    int key;
    Node*next, *prev;

    Node(int val, int key){
        this->val = val;
        this->key = key;
        this->next = NULL;
        this->prev = NULL;
    }
};

class LRUCache {
    int cap;
    int size;
    Node* head, *tail;
    map<int, Node*>mp;

public:
    LRUCache(int capacity) {
        this->cap = capacity;
        this->size =0;
        head = new Node(-1,-1);
        tail = new Node(-1,-1);
        head->next = tail;
        tail->prev = head;
    }
    
    void erase(Node*curr){
        mp.erase(curr->key);
        Node*prev = curr->prev;
        Node*next = curr->next;
        prev->next = next;
        next->prev = prev;
        size--;
    }
    Node* append(int key, int val){
        Node*curr = new Node(val, key);
        Node*next = head->next;
        curr->prev = head;
        head->next = curr;
        next->prev = curr;
        curr->next = next;
        size++;
        return curr;
    }


    int get(int key) {
        if(mp.find(key)==mp.end()) return -1;
        Node*curr = mp[key];
        int value = curr->val;
        erase(curr);
        mp[key] = append(key, value);
        return value;
    }
    
    void put(int key, int value) {
        if(mp.find(key)==mp.end()){
            if(size==cap){
                erase(tail->prev);
            }
            mp[key] = append(key, value);
        }
        else{
            Node*curr = mp[key];
            erase(curr);
            mp[key] = append(key,value);
        }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */